app/gui/controller/double-representation/src/node.rs

```diff
@@ -330,9 +348,14 @@ impl MainLine {
     pub fn set_expression(&mut self, expression: Ast) {
         let id = self.id();
         match self {
-            MainLine::Binding { ref mut infix } =>
-                infix.update_shape(|infix| infix.rarg = expression),
-            MainLine::Expression { ref mut ast } => *ast = expression,
+            MainLine::Binding { ref mut infix, ref mut macros_info } => {
+                *macros_info = MacrosInfo::from_ast(&expression);
+                infix.update_shape(|infix| infix.rarg = expression);
+            }
+            MainLine::Expression { ref mut ast, ref mut macros_info } => {
+                *macros_info = MacrosInfo::from_ast(&expression);
+                *ast = expression
+            }
```

Comment:
AFAIK, in most cases, we want to set visible expression only here. Now, I think editing a node with skip turned out will remove skip.

I'd consider renaming visible expression to just expression, and remove the old expression getter altogether. And the set_expression should just preserve skip and freeze macros.

Reply:
Yep, that makes sense. Changed as proposed.